libgeopm/src/DGEMMModelRegion.cpp

```diff
@@ -71,43 +73,76 @@ namespace geopm
 
     void DGEMMModelRegion::cleanup(void)
     {
-        free(m_matrix_c);
-        free(m_matrix_b);
-        free(m_matrix_a);
+        if (m_matrix_a) {
+            free(m_matrix_a);
+            m_matrix_a = NULL;
+        }
+        if (m_matrix_b) {
+            free(m_matrix_b);
+            m_matrix_b = NULL;
+        }
+        if (m_matrix_c) {
+            free(m_matrix_c);
+            m_matrix_c = NULL;
+        }
+    }
+
+    void DGEMMModelRegion::num_progress_updates(double big_o_in)
+    {
+        m_num_progress_updates = (uint64_t)(100.0 * big_o_in);
+        (void)geopm_tprof_init(m_num_progress_updates);
     }
 
     void DGEMMModelRegion::big_o(double big_o_in)
     {
         if (m_big_o && m_big_o != big_o_in) {
             cleanup();
         }
-
         geopm_prof_region("geopm_dgemm_model_region_startup", GEOPM_REGION_HINT_IGNORE, &m_start_rid);
         geopm_prof_enter(m_start_rid);
 
         num_progress_updates(big_o_in);
 
-        m_matrix_size = (int)pow(4e9 * big_o_in / m_num_progress_updates, 1.0/3.0);
         if (big_o_in && m_big_o != big_o_in) {
-            size_t mem_size = sizeof(double) * (m_matrix_size * (m_matrix_size + m_pad_size));
-            int err = posix_memalign((void **)&m_matrix_a, m_pad_size, mem_size);
-            if (!err) {
-                err = posix_memalign((void **)&m_matrix_b, m_pad_size, mem_size);
+            // Allocate A: M x K
+            size_t mem_size_a = sizeof(double) * (m_matrix_m_size + m_pad_size) * m_matrix_k_size;
+            int err = posix_memalign((void **)&m_matrix_a, m_pad_size, mem_size_a);
+            if (err) {
+                throw Exception("DGEMMModelRegion::big_o(): posix_memalign() failed",
+                                err, __FILE__, __LINE__);
             }
-            if (!err) {
-                err = posix_memalign((void **)&m_matrix_c, m_pad_size, mem_size);
+#ifdef GEOPM_ENABLE_OMPT
+#pragma omp parallel for
+#endif
+            for (size_t i = 0; i < mem_size_a / sizeof(double); ++i) {
+                m_matrix_a[i] = 2.0 * i;
             }
+            // Allocate B: K x N
+            size_t mem_size_b = sizeof(double) * (m_matrix_k_size + m_pad_size) * m_matrix_n_size;
+            err = posix_memalign((void **)&m_matrix_b, m_pad_size, mem_size_b);
             if (err) {
+                free(m_matrix_a);
+                m_matrix_a = NULL;
```

Comment:
Do we need to check if m_matrix_a is already NULL to prevent a double free, in case someone is catching the previous exception? Same below for matrix a/b on 139/140.

Reply:
Don't think that control flow is possible.